bindings/matrix-sdk-ffi/src/encryption.rs

```diff
@@ -0,0 +1,328 @@
+use std::sync::Arc;
+
+use futures_util::StreamExt;
+use matrix_sdk::encryption::{backups, recovery};
+use thiserror::Error;
+use zeroize::Zeroize;
+
+use super::RUNTIME;
+use crate::{error::ClientError, task_handle::TaskHandle};
+
+#[derive(uniffi::Object)]
+pub struct Encryption {
+    inner: matrix_sdk::encryption::Encryption,
+}
+
+impl From<matrix_sdk::encryption::Encryption> for Encryption {
+    fn from(value: matrix_sdk::encryption::Encryption) -> Self {
+        Self { inner: value }
+    }
+}
+
+#[uniffi::export(callback_interface)]
+pub trait BackupStateListener: Sync + Send {
+    fn on_update(&self, status: BackupState);
+}
+
+#[uniffi::export(callback_interface)]
+pub trait BackupSteadyStateListener: Sync + Send {
+    fn on_update(&self, status: BackupUploadState);
+}
+
+#[uniffi::export(callback_interface)]
+pub trait RecoveryStateListener: Sync + Send {
+    fn on_update(&self, status: RecoveryState);
+}
+
+#[derive(uniffi::Enum)]
+pub enum BackupUploadState {
+    Waiting,
+    Uploading { backed_up_count: u32, total_count: u32 },
+    Error,
+    Done,
+}
+
+#[derive(Debug, Error)]
+pub enum SteadyStateError {
+    #[error("The backup got disabled while waiting for the room keys to be uploaded.")]
+    BackupDisabled,
+    #[error("There was a connection error.")]
+    Connection,
+    #[error("We couldn't read status updates from the upload task quickly enough.")]
+    Lagged,
+}
+
+#[derive(Debug, Error)]
+pub enum RecoveryError {
+    /// A backup already exists on the homeserver, the recovery subsystem does
+    /// not allow backups to be overwritten, disable recovery first.
+    #[error(
+        "A backup already exists on the homeserver and the method does not allow to overwrite it"
+    )]
+    BackupExistsOnServer,
+
+    /// A typical SDK error.
+    #[error(transparent)]
+    Client { source: crate::ClientError },
+
+    /// Error in the secret storage subsystem.
+    #[error("Error in the secret-storage subsystem: {message}")]
```

Comment:
Since this is not a flat error, the derived `std::error::Error` is likely never used anywhere (I guess it doesn't really hurt either though).

Reply:
I'll leave it as is then.